tests/test_irdl.py

```diff
@@ -0,0 +1,243 @@
+        else:
+            raise Exception("Wrong argument passed to BoolAttr.")
+
+    @staticmethod
+    def print_parameter(data: bool, printer: Printer):
+        printer.print_string(str(data))
+
+
+@irdl_attr_definition
+class IntData(Data[int]):
+    """An attribute holding an integer value."""
+    name = "int"
+
+    @staticmethod
+    def parse_parameter(parser: Parser) -> int:
+        return parser.parse_int_literal()
+
+    @staticmethod
+    def print_parameter(data: int, printer: Printer):
+        printer.print_string(str(data))
+
+
+@irdl_attr_definition
+class DoubleParamAttr(ParametrizedAttribute):
+    """An attribute with two unbounded attribute parameters."""
+    name = "param"
+
+    param1: ParameterDef[Attribute]
+    param2: ParameterDef[Attribute]
+
+
+def test_eq_attr_verify():
+    """Check that an EqAttrConstraint verifies the expected attribute"""
+    bool_true = BoolData(True)
+    eq_true_constraint = EqAttrConstraint(bool_true)
+    eq_true_constraint.verify(bool_true)
+
+
+def test_eq_attr_verify_wrong_parameters_fail():
+    """
+    Check that an EqAttrConstraint fails to verify an attribute with different
+    parameters.
+    """
+    bool_true = BoolData(True)
+    bool_false = BoolData(False)
+    eq_true_constraint = EqAttrConstraint(bool_true)
+    with pytest.raises(VerifyException) as e:
+        eq_true_constraint.verify(bool_false)
+    assert e.value.args[0] == (
+        f"Expected attribute {bool_true} but got {bool_false}")
+
+
+def test_eq_attr_verify_wrong_base_fail():
+    """
+    Check that an EqAttrConstraint fails to verify an attribute with a
+    different base attribute.
+    """
+    bool_true = BoolData(True)
+    int_zero = IntData(0)
+    eq_true_constraint = EqAttrConstraint(bool_true)
+    with pytest.raises(VerifyException) as e:
+        eq_true_constraint.verify(int_zero)
+    assert e.value.args[0] == (
+        f"Expected attribute {bool_true} but got {int_zero}")
+
+
+def test_base_attr_verify():
+    """
+    Check that a BaseAttr constraint verifies an attribute with the expected
+    base attribute.
+    """
+    eq_true_constraint = BaseAttr(BoolData)
+    eq_true_constraint.verify(BoolData(True))
+    eq_true_constraint.verify(BoolData(False))
+
+
+def test_base_attr_verify_wrong_base_fail():
+    """
+    Check that a BaseAttr constraint fails to verify an attribute with a
+    different base attribute.
+    """
+    eq_true_constraint = BaseAttr(BoolData)
+    int_zero = IntData(0)
+    with pytest.raises(VerifyException) as e:
+        eq_true_constraint.verify(int_zero)
+    assert e.value.args[0] == (
+        f"{int_zero} should be of base attribute {BoolData.name}")
+
+
+def test_any_attr_verify():
+    """Check that an AnyAttr verifies any attribute."""
+    any_constraint = AnyAttr()
+    any_constraint.verify(BoolData(True))
+    any_constraint.verify(BoolData(False))
+    any_constraint.verify(IntData(0))
+
+
+def test_anyof_verify():
```

Comment:
It might be more intuitive to test the `AnyOf` in the same way as `AllOf`. So through the use of integer constraints.

Reply:
Good idea, done1